File: 2025/04 April/17 Minimum Weight Cycle/main.cpp
// Link: https://www.geeksforgeeks.org/problems/minimum-weight-cycle/1

#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int findMinCycle(int V, vector<vector<int>> &edges)
    {
        vector<vector<pair<int, int>>> graph(V);
        for (const auto &edge : edges)
        {
            int u = edge[0], v = edge[1], w = edge[2];
            graph[u].emplace_back(v, w);
            graph[v].emplace_back(u, w);
        }

        const int INF = numeric_limits<int>::max();
        int min_cycle = INF;

        for (const auto &edge : edges)
        {
            int u = edge[0], v = edge[1], w = edge[2];
            graph[u].erase(remove(graph[u].begin(), graph[u].end(), make_pair(v, w)), graph[u].end());
            graph[v].erase(remove(graph[v].begin(), graph[v].end(), make_pair(u, w)), graph[v].end());

            vector<int> dist(V, INF);
            vector<bool> visited(V, false);
            dist[u] = 0;

            for (int i = 0; i < V; ++i)
            {
                int min_dist = INF;
                int min_node = -1;
                for (int j = 0; j < V; ++j)
                {
                    if (!visited[j] && dist[j] < min_dist)
                    {
                        min_dist = dist[j];
                        min_node = j;
                    }
                }

                if (min_node == -1)
                {
                    break;
                }

                visited[min_node] = true;

                for (const auto &neighbor : graph[min_node])
                {
                    int neighbor_node = neighbor.first;
                    int weight = neighbor.second;
                    if (!visited[neighbor_node] && dist[neighbor_node] > dist[min_node] + weight)
                    {
                        dist[neighbor_node] = dist[min_node] + weight;
                    }
                }
            }

            if (dist[v] != INF)
            {
                min_cycle = min(min_cycle, dist[v] + w);
            }

            graph[u].emplace_back(v, w);
            graph[v].emplace_back(u, w);
        }

        return min_cycle != INF ? min_cycle : -1;
    }
};

int main()
{
    Solution obj;
    int V = 5;
    vector<vector<int>> edges = {{0, 1, 2}, {1, 2, 2}, {1, 3, 1}, {1, 4, 1}, {0, 4, 3}, {2, 3, 4}};
    cout << obj.findMinCycle(V, edges) << endl;

    V = 5;
    edges = {
        {0, 1, 3},
        {1, 2, 2},
        {0, 4, 1},
        {1, 4, 2},
        {1, 3, 1},
        {3, 4, 2},
        {2, 3, 3},
    };
    cout << obj.findMinCycle(V, edges) << endl;

    return 0;
}

File: 2025/04 April/17 Minimum Weight Cycle/main.py
# Link: https://www.geeksforgeeks.org/problems/minimum-weight-cycle/1


class Solution:
    def findMinCycle(self, V, edges):
        graph = [[] for _ in range(V)]
        for u, v, w in edges:
            graph[u].append((v, w))
            graph[v].append((u, w))

        INF = float("inf")
        min_cycle = INF

        for u, v, w in edges:
            graph[u] = [pair for pair in graph[u] if pair != (v, w)]
            graph[v] = [pair for pair in graph[v] if pair != (u, w)]

            dist = [INF] * V
            visited = [False] * V
            dist[u] = 0

            for _ in range(V):
                min_dist = INF
                min_node = -1
                for i in range(V):
                    if not visited[i] and dist[i] < min_dist:
                        min_dist = dist[i]
                        min_node = i

                if min_node == -1:
                    break

                visited[min_node] = True

                for neighbor, weight in graph[min_node]:
                    if (
                        not visited[neighbor]
                        and dist[neighbor] > dist[min_node] + weight
                    ):
                        dist[neighbor] = dist[min_node] + weight

            if dist[v] != INF:
                min_cycle = min(min_cycle, dist[v] + w)

            graph[u].append((v, w))
            graph[v].append((u, w))

        return min_cycle if min_cycle != INF else -1


if __name__ == "__main__":
    obj = Solution()
    V = 5
    edges = [[0, 1, 2], [1, 2, 2], [1, 3, 1], [1, 4, 1], [0, 4, 3], [2, 3, 4]]
    print(obj.findMinCycle(V, edges))

    V = 5
    edges = [
        [0, 1, 3],
        [1, 2, 2],
        [0, 4, 1],
        [1, 4, 2],
        [1, 3, 1],
        [3, 4, 2],
        [2, 3, 3],
    ]
    print(obj.findMinCycle(V, edges))
